src/interpreter/mod.rs

```diff
@@ -65,7 +65,13 @@ impl Interpreter {
         Ok(())
     }
 
-    fn eval_enum_stmt(&mut self, stmt: &EnumDecl) -> StmtVal {
+    fn eval_enum_stmt(&mut self, stmt: &EnumDecl) -> Result<StmtVal> {
+        if self.env.borrow().contains_key(&stmt.name.lexeme) {
```

Comment:
It would be better not to burden interpreter with such checks. Lets move this code directly to `define_enum`

Reply:
The other option is to check before the `struct_`/`enum_` is inserted to the environment in `define_enum` or `define_struct` as you've mentioned. However, in case it results in a name collision, the declaration would be have been processed(into a `enum_` or a `struct_`) unnecessarily. What would you suggest in this case?

Reply:
that's a good point, actually. I think we would go your way as it is indeed evades the unnecessary interpreter work. 

Let's rename it to something more specific `contains_key` -> `is_defined`. This way the public api of `Env` would look more appropriate.

Reply:
How about `has_definition`?